src/spac/visualization.py

```diff
@@ -7,6 +7,116 @@
 import math
 import matplotlib.pyplot as plt
 from matplotlib.colors import ListedColormap, BoundaryNorm
+from spac.utils import check_table, check_annotation, check_feature
+
+
+def boxplot(adata, annotation=None, second_annotation=None,
+            layer=None, ax=None, features=None, **kwargs):
+    """
+    Create a boxplot visualization of the features in the passed adata object.
+    This function offers flexibility in how the boxplots are displayed,
+    based on the arguments provided.
+
+    Parameters
+    ----------
+    adata : anndata.AnnData
+        The AnnData object.
+
+    annotation : str, optional
+        Annotation to determine if separate plots are needed for every label.
+
+    second_annotation : str, optional
+        Second annotation to further divide the data.
+
+    layer : str, optional
+        The name of the matrix layer to use. If not provided,
+        uses the main data matrix adata.X.
+
+    ax : matplotlib.axes.Axes, optional
+        An existing Axes object to draw the plot onto, optional.
+
+    features : list, optional
+        List of feature names to be plotted.
+        If not provided, all features will be plotted.
+
+    **kwargs
+        Additional arguments to pass to seaborn.boxplot.
+
+    Returns
+    -------
+    fig, ax : matplotlib.figure.Figure, matplotlib.axes.Axes
+        The created figure and axes for the plot.
+
+    Examples
+    --------
+    - Multiple features boxplot: boxplot(adata, features=['GeneA','GeneB'])
+    - Boxplot grouped by a single annotation:
+      boxplot(adata, features=['GeneA'], annotation='cell_type')
+    - Nested grouping by two annotations: boxplot(adata, features=['GeneA'],
+      annotation='cell_type', second_annotation='treatment')
+    """
+
+    # Use utility functions to check inputs
+    check_table(adata, tables=layer)
+    check_annotation(adata, annotations=annotation)
+    if second_annotation:
+        check_annotation(adata, annotations=second_annotation)
+    check_feature(adata, features=features)
+
+    # Validate ax instance
+    if ax and not isinstance(ax, plt.Axes):

```

Comment:
@fangliu117, if that's a common error we want to check, let's move it to the utils.py module.

Reply:
Thank you! This is a good point. In the current implementation, second_annotation is only considered if annotation is also provided. This is because the second_annotation is used to add an additional level of nesting within the primary categories defined by annotation. I modified the code, and in this case, second_annotation would function as the primary grouping variable, rather than adding a level of nesting.  
So the function will create a boxplot grouped by second_annotation if annotation is None but second_annotation is provided.